### ISSUE TITLE
Feature Request: Ability to write one track at a time on Foundry

### ISSUE THREAD

**Issue by fbeecher** [MarcBoule/ImpromptuModular#28]
(This FR comes from a discussion on the VCV Rack Community site: https://community.vcvrack.com/t/impromptu-0-6-13-update-new-sequencer-foundry/1330)

**Current State**
When a trigger is sent to the WRITE input, Foundry writes all eight voltages coming into CV A-D and CV2 A-D to the current step. This makes it impossible to program each of the four tracks independently only using external controllers. Doing so requires that only a single voltage be patched into Foundry, making it so that re-patching in the Rack UI is necessary to record subsequent tracks.

**Desired State**
Ideally, users should be able to keep each CV input patched while controlling which track(s?) a voltage is written to when a trigger signal is received at the WRITE input.

There are potentially many ways of doing this, including using utility modules to alter signals going into the CV inputs and specifying voltages which would be interpreted as "don't write to this track." Since there are so many ways to accomplish this, I'll articulate a few of my personal priorities. These are not based on any sort of user research whatsoever.

**Priorities**
1. Being able to write a single voltage from a single CV input to one step on a single track.
2. Being able to visually see which single CV input will be written to
3. Being able to select the track(s) to write to via CV and/or gate signals
4. Accomplish the above goals without using additional modules
5. Being able to write more than one but less than eight voltages at a time
6. Being able to visually see which CV inputs will be written to

**Comment by MarcBoule** [comment]
Thanks for posting this here, it's perfect. It's going to be a challenge, but I'll see what can be done while keeping the interface relatively intact.

Here's one idea I had to help manage how the CVs get written: I could modify the TRACK cv input such that instead of a 0-10V input mapping to the {A, B, C, D} track selection, it could map as follows:
{A, B, C, D, A*, B*, C*, D*}. I could then change the code so that writing to all tracks is only done when the star (ALL) is on. And we would be able to control all this with the TRACK cv input, so that could solve part of the problem in point 1 without any UI changes. Next would be how to control whether we want CV, or CV2, or both to be written. If were to add a single LED button next to the CV IN group of jack, and another next to the CV2 IN group of jacks, that would allow us to disable either one of those two groups of inputs, but would require a mouse click. So also implicit in your list of priorities, I imagine, is to be able to make this choice, or ultimately the more detailed choices in point 4 controllable though CV, as opposed to adding push buttons on the module, correct?

I will have to think more about this, but thanks for followup.

**Comment by fbeecher** [comment]
Hi Marc. I've thought about this a little bit and mocked up an example UI. Basically, it relies on the track select buttons to indicate which single track to write to. It also allows users to select whether they want to write to all steps or just one using the same UI pattern you use for switching between CV2, P, and R. (This may interfere with what ALL is currently used for, so spend some brain cycles on this one please.)

To make room for this I changed CLK RES/DLY to CLK EDIT. Dunno if this still makes sense. :)

I've also added some indicator LEDs next to the CV A-D inputs to indicate which is/are selected.

This does not yet solve for the writing to CV2 problem. I just want to get this idea out there to help the process along and see where we go from here.

Thanks!

![foundry_ui-fb](https://user-images.githubusercontent.com/36204932/51687009-34fd4300-1fb7-11e9-8f9c-16bcb8918035.png)

**Comment by MarcBoule** [comment]
Thanks for taking the time to do the mock up. For the LEDs near the jacks, I would probably go with smaller ones like Fundamental 8-vert (maybe not as small, but in that style). But the idea seems good.

![image](https://user-images.githubusercontent.com/33604154/51688642-07b69200-1fc3-11e9-83cb-230ce9160572.png)

Not sure I totally understand your solution though. In what you propose, it looks like we can only select ALL for writing via the CV inputs, but not for other normal operations on the module itself, like turning on gates, etc. The functionality of the current ALL button seems to be lost no?. Unless you provide for the fact that the "-" and "+" buttons will also cycle the tracks like I proposed in my post above regarding the TRACK CV input, i.e: A, B, C, D, A*, B*, C*, D*. If we assume we are programming the sequencer with external controllers, then my idea with the TRACK cv input seems to do the same thing as your proposal but without any UI changes.

Or by WRT ALL/1 do you mean that it controls all the writing operations, whether by CV input or by pressing buttons on the module itslef?

**Comment by fbeecher** [comment]
You press the WRT button to toggle between CVs writing to ALL tracks or 1 track. If you select 1 track, the track that gets written to is the one selected above using the -/+ buttons. And yes, it does kind of mess with the current ALL functionality.

I think I'm seeing my own bias in my solution vs. yours. :) For me, one track at a time is the primary use case. But it might be exactly the opposite for most users. I'm seeing this as a primary function. If you implement your solution, the goal can still be accomplished but it's more a "neat trick" that Foundry can do. 

The thing I like about your solution, though, is that I can set up my controller to have N different buttons that all send different values of the same MIDI CC. On my LaunchControl XL in particular that would work really well. 

One question, what does A* mean exactly? Is that CV2? I thought it meant "this step on all tracks" or something like that. Would it make sense for the TRACK CV input to be more like A, B, C, D, A2, B2, C2, D2 to handle CV2? Another option would be A, A2, B, B2, C, C2, D, D2.

**Comment by MarcBoule** [comment]
N* refers to the case when we are on track N, which we can go to using the +- buttons, and the * means that we are in ALL mode (which we can toggle with the ALL button). N* means that we are currently seeing track N in the displays, but that any change we make (i.e. change a note, a gate, write CVs via the inputs, etc) will be done in all tracks (*) in parallel.

For determining wheter to write CV, CV2 or both, I thought of a little hack that would again not require any UI changes: each time we cycle through the CV2 state using the CV2/p/r button, then it would cycle through the writing choices: CV, CV2 or both. This would be visible in the leds near the jacks, as we would see which ones are on or off. It would also visually show the effect of toggling the ALL button: it would toggle between all 4 leds near a group of CV(2) inputs being active, for example, and just one of those jacks being active. If this is acceptable, I may try it. I think it's a good compromise between added functionality, and not adding buttons.  

If you want to try the new current version though, with the new TRACK cv input behavior, just let me know your platform and I'll send you a build. Really appreciate the discussion! :-)

**Comment by MarcBoule** [comment]
> Would it make sense for the TRACK CV input to be more like A, B, C, D, A2, B2, C2, D2 to handle CV2? Another option would be A, A2, B, B2, C, C2, D, D2. Another option could be the latter plus A*, B*, C*, D* tacked on to the end.

I would need to add another character in the display for somehting like A2*, but that would be one way to do it. We would need to distinguish between three possibilities though: write CV only, write CV2 only, write BOTH. 

My CV2 button cycling idea requires nothing more to be added, but I agree it's not easy to grasp. I think I'll go ahead and code it so you can try it, that might be best.

This is what pressing the CV2 button in repetition would do:
```
CV2   write CV only
p     ""
r     ""
CV2   write CV2 only
p     ""
r     ""
CV2   write both
p     ""
r     ""
```

If this is unclear, trying it out will hopefully make it clearer. Yes, putting the zip file in the plugins folder is all that needs to be done. I'll send it when I finish the code and I'm back at my mac either tonight or tomorrow night. Thanks!

**Comment by MarcBoule** [comment]
[ImpromptuModular-0.6.13-mac.zip](https://github.com/MarcBoule/ImpromptuModular/files/2798512/ImpromptuModular-0.6.13-mac.zip)

Here's a MAC build so you can test the new writing modes. Instead of the cryptic trick in the comment above, I decided to put a button in the expansion panel (layout not finalized) to toggle between these 3 possibilities:

- write both
- write CV only
- write CV2 only

This should cover most of the numbered points you identified, point 5 is not handled, but that would really mean widening the panel so not sure I will go there. Please let me know what you think about this version. Thanks!

**Comment by fbeecher** [comment]
That. Is. So. Much. FUN!

Honestly, I think that even with a simple little patch (4xFMOP, 4xQuantizer, Mixer, Delay) that is the most fun I've ever had with Rack. DANG! Track selection via CC works really slick in concert with this new functionality too.

I do have a couple of thoughts. I've listed them in order of priority below.

1. The button/LED pattern works, but a good label and placement will be critical. Its current location makes it seem like it's associated only with CV2
2. Optional: Change CV write mode via CV
3. Optional: Toggle Autostep via Gate
4. Optional: Change probability and slide time via CV
5. Optional: Write current step voltage to _all steps in the current track_ (CV and CV2)
6. Optional: Marc gets a job at Arturia and implements Item 5 on the Beatstep Pro. :)

The only thing that really needs to be addressed is the button label. The rest is just scope creep. :) If you choose to implement any of that, though, I'm guessing either a second expander or an expanded expander would be necessary. They could potentially be split up by gates & CVs, or by the tasks that they support (e.g., programing CV2 vs programming Foundry via hardware).

[Foundry Experiment.vcv.zip](https://github.com/MarcBoule/ImpromptuModular/files/2799185/Foundry.Experiment.vcv.zip)

**Comment by MarcBoule** [comment]
Well I like the sound of number 6! :-) 

For number 5 though, I don't know if you've used the SEL button yet on Foundry, but if you move to the first step, click SEL, and put the switch in END position, you should be able to write into all steps of the sequence in one shot. 

You're correct that it would have to be another panel, and such options are indeed endless, but at least we have improved things a bit. I'm having lots of trouble getting that last switch done which, as you mention, is important to properly place and label. Here's my paint mockup, just to show you what I'm thinking in case you have any ideas. I think it should be in the expansion panel, since when the panel is inactive, it reverts to always activating CV for writing and the switch has no use. 

![image](https://user-images.githubusercontent.com/33604154/51792102-b81ac680-217a-11e9-9d24-306815631c4e.png)

**Comment by NRG70** [comment]
interesting conversation and improvements guys:)

**Comment by fbeecher** [comment]
Thanks for the SEL/END explanation! I will definitely put that to good use.

Regarding having the button on the expansion panel, I absolutely agree. I do have a few thoughts on how that panel should be organized. See attached mockup.

Basically, I tried to go from the "biggest" parameters (selecting sequences, tracks) to the "smallest" (individual step parameters), left to right. I've also changed the CV write mode button to a switch, as that already exists in your visual language and communicates the situation a bit more clearly. One suggestion I did NOT mockup... maybe the indicator LEDs should be red? Most other LEDs in your stuff are red, and also the contrast against the gray is a little better. I'm not sure about this tho. While it's good for contrast, you clearly use green on the RUN button to indicate an "on" status.

What this doesn't do is use all the vertical space. However, this does mean that space exists to CV-ify (technical term, obviously :) additional parameters such as gate mode, sequence length, etc.

I'd love to hear your thoughts!

![foundry_ui-fb_v2](https://user-images.githubusercontent.com/36204932/51928874-8729d400-23bc-11e9-890e-a5a735e90a00.png)

**Comment by MarcBoule** [comment]
I thought about making the CV IN lights red also, but then figured the same thing you said, about green being "on". In the case of the run switch, it has a different format, and green goes well there, but after reading your comment and thinking about it more, I think it would be better if the CV IN lights were indeed red, since all other such lights are red (for example, an "on" gate is red, same with slide and tied, and even attach is red when "on", for that matter). I just tried it now with red, and I think it is more hamonious with the overall feel of the module.

For the layout, I agree that the way I have it now it not ideal, and if/when I decide to add more CV inputs to the expansion panel I will surely do a new layout an improve that aspect. At least now though, users can press the button and see what happens in the lights, which are more visible now that they are red, so I think I'll leave it like this for now. Since the expansion panel does not automatically nudge modules to the right when we activate it (the exp panel will overlap modules to the right if we don't make space), I want to keep it as narrow as possible.

I also moved the more general inputs to the top like you suggested, which makes sense (more general to more specific, indeed).

Here's the new look:

![image](https://user-images.githubusercontent.com/33604154/51931468-de36a580-23ca-11e9-9a22-b332cefcf46e.png)

This should be hitting the plugin manager in a few days. Cheers!
